lib/extensions/src/lib/components/dynamic-component/dynamic.component.ts

```diff
@@ -19,49 +19,98 @@ import {
     Component,
     Input,
     ComponentRef,
-    OnInit,
     ComponentFactoryResolver,
     ViewChild,
     ViewContainerRef,
-    OnDestroy
+    OnDestroy,
+    OnChanges,
+    SimpleChanges
 } from '@angular/core';
 import { ExtensionService } from '../../services/extension.service';
+import { ExtensionComponent } from '../../services/component-register.service';
 
+// cSpell:words lifecycle
 @Component({
     selector: 'adf-dynamic-component',
     template: `<div #content></div>`
 })
-export class DynamicExtensionComponent implements OnInit, OnDestroy {
+export class DynamicExtensionComponent implements OnChanges, OnDestroy {
     @ViewChild('content', { read: ViewContainerRef })
     content: ViewContainerRef;
 
     @Input() id: string;
+    @Input() data: any;
 
-    private componentRef: ComponentRef<any>;
+    private componentRef: ComponentRef<ExtensionComponent>;
+    private loaded: boolean = false;
 
-    constructor(
-        private extensions: ExtensionService,
-        private componentFactoryResolver: ComponentFactoryResolver
-    ) {}
+    constructor(private extensions: ExtensionService, private componentFactoryResolver: ComponentFactoryResolver) {
+        const dynamicLifeCycleMethods = [
+            'ngOnInit',
+            'ngDoCheck',
+            'ngAfterContentInit',
+            'ngAfterContentChecked',
+            'ngAfterViewInit',
+            'ngAfterViewChecked'
+        ];
 
-    ngOnInit() {
-        const componentType = this.extensions.getComponentById(this.id);
+        dynamicLifeCycleMethods.forEach((method) => {
+            this[method] = this.proxy.bind(this, method);
+        });
+    }
+
+    ngOnChanges(changes: SimpleChanges) {
+        if (!this.loaded) {
+            this.loadComponent();
+            this.loaded = true;
+        }
+
+        if (changes.data) {
+            this.data = changes.data.currentValue;
+        }
+
+        this.updateInstance();
+        this.proxy('ngOnChanges', changes);
+    }
+
+    private loadComponent() {
+        const componentType = this.extensions.getComponentById<ExtensionComponent>(this.id);
         if (componentType) {
             const factory = this.componentFactoryResolver.resolveComponentFactory(
                 componentType
             );
             if (factory) {
                 this.content.clear();
                 this.componentRef = this.content.createComponent(factory, 0);
-                // this.setupWidget(this.componentRef);
             }
         }
     }
 
     ngOnDestroy() {
-        if (this.componentRef) {
+        if (this.componentCreated()) {
+            this.proxy('ngOnDestroy');
             this.componentRef.destroy();
             this.componentRef = null;
         }
     }
+
+    private updateInstance() {
+        if (this.componentCreated()) {
+            this.componentRef.instance.data = this.data;
+        }
+    }
+
+    private proxy(lifecycleMethod, ...args) {
```

Comment:
When created dynamically, child component will already have own lifecycle methods, like onInit, of onDestroy. You may be calling them 2 times now (might cause errors with double destroy or double init)

Reply:
I suggest checking manually at runtime, by putting console.log in the component (to avoid components not being ready during test run). Angular should wire those with all life-cycle stuff properly. The only issue I had is propagating custom data changes down to the child ([example](https://github.com/Alfresco/alfresco-content-app/blob/master/src/app/components/common/dynamic-column/dynamic-column.component.ts#L67)) to have like `<dynamic-component [context]="node">` and then `node` value being passed to constructed child each time it changes.

Reply:
Ok, thanks, I'm checking it now, and try to cover it with better tests.

Reply:
Tested it in both the testing environment and browser. You are right. Testing it in browser, every lifecycle method of the dynamically created component was invoked properly without the need of patching it. However the testing environment for some reason can't do the same [which is very misleading]. So I remove those patches and tests.

I left the patch only for the ngOnChanges as you suggested, and the ngOnDestroy will be called because of the componentRef.destroy method.